File: findAnagrams.py
#Approach- Sliding window with Hashmap 

from collections import Counter
def findAnagrams(s,p):             # TC O(Ns+Np) // SC O(26) --> O(1)
	ns, np = len(s), len(p)
	if ns < np:
	    return []

	p_count = Counter(p)
	s_count = Counter()

	output = []
	# sliding window on the string s NOTE- but is actually on hashmap so need of sliding window by 1
	for i in range(ns):
	    # add one more letter 
	    # on the right side of the window
	    s_count[s[i]] += 1
	    # remove one letter 
	    # from the left side of the window
	    if i >= np:
	        if s_count[s[i - np]] == 1:
	            del s_count[s[i - np]]
	        else:
	            s_count[s[i - np]] -= 1
	    # compare array in the sliding window
	    # with the reference array
	    if p_count == s_count:
	        output.append(i - np + 1)

	return output


#Approach- Sliding window with Array

def findAnagrams1(s,p):             # TC O(Ns+Np) // SC O(26) --> O(1)
	ns, np = len(s), len(p)
	if ns < np:
	    return []

	p_count, s_count = [0] * 26, [0] * 26
	# build reference array using string p
	for ch in p:
	    p_count[ord(ch) - ord('a')] += 1

	output = []
	# sliding window on the string s
	for i in range(ns):
	    # add one more letter 
	    # on the right side of the window
	    s_count[ord(s[i]) - ord('a')] += 1
	    # remove one letter 
	    # from the left side of the window
	    if i >= np:
	        s_count[ord(s[i - np]) - ord('a')] -= 1
	    # compare array in the sliding window
	    # with the reference array
	    if p_count == s_count:
	        output.append(i - np + 1)

	return output

print(findAnagrams("cbaebabacd","abc"))
print(findAnagrams1("abab","ab"))

File: findAnagrams.java
import java.util.*;
// Approach 1- Sliding window with HashMaps 
class Solution { // TC O(Ns+Np) SC O(1)
  public List<Integer> findAnagrams(String s, String p) {
    int ns = s.length(), np = p.length();
    if (ns < np) return new ArrayList<>();

    Map<Character, Integer> pCount = new HashMap<>();
    Map<Character, Integer> sCount = new HashMap<>();
    // build reference hashmap using string p
    for(char c: p.toCharArray()){
            pCount.put(c,pCount.getOrDefault(c,0)+1);
        }

    List<Integer> output = new ArrayList<>();
    // sliding window on the string s
    for (int i = 0; i < ns; ++i) {
      // add one more letter 
      // on the right side of the window
      char ch = s.charAt(i);
      sCount.put(ch,sCount.getOrDefault(ch,0)+1);
      // remove one letter 
      // from the left side of the window
      if (i >= np) {
        ch = s.charAt(i - np);
        if (sCount.get(ch) == 1) {
          sCount.remove(ch);
        }
        else {
          sCount.put(ch, sCount.get(ch) - 1);
        }
      }
      // compare hashmap in the sliding window
      // with the reference hashmap
      if (pCount.equals(sCount)) {
        output.add(i - np + 1);
      }
    }
    return output;
  }
}

// Approach 2- Sliding window with arrays 
class Solution { // TC O(Ns+Np) SC O(1)
  public List<Integer> findAnagrams(String s, String p) {
    int ns = s.length(), np = p.length();
    if (ns < np) return new ArrayList();

    int [] pCount = new int[26];
    int [] sCount = new int[26];
    // build reference array using string p
    for (char ch : p.toCharArray()) {
      pCount[ch - 'a']++;
    }

    List<Integer> output = new ArrayList();
    // sliding window on the string s
    for (int i = 0; i < ns; ++i) {
      // add one more letter 
      // on the right side of the window
      sCount[s.charAt(i) - 'a']++;
      // remove one letter 
      // from the left side of the window
      if (i >= np) {
        sCount[s.charAt(i - np) - 'a']--;
      }
      // compare array in the sliding window
      // with the reference array
      if (Arrays.equals(pCount, sCount)) {
        output.add(i - np + 1);
      }
    }
    return output;
  }
}
class findAnagrams{
  public static void main(String[] args) {
    String s="cbaebabacd";
    String p="abc";
    Solution x= new Solution();
    System.out.println(x.findAnagrams(s,p));
  }
}